packages/react-magma-dom/src/hooks/usePagination.ts

```diff
@@ -0,0 +1,204 @@
+import React from 'react';
+import { I18nContext } from '../i18n';
+import { useControlled } from './useControlled';
+
+export interface UsePaginationProps {
+  /**
+   * The total number of Pagination buttons
+   * @default 1
+   */
+  count?: number;
+  /**
+   * Page selected by default when the component is uncontrolled
+   */
+  defaultPage?: number;
+  /**
+   * If true, disables all of the Pagination buttons
+   * @default false
+   */
+  disabled?: boolean;
+  /**
+   * If true, hides the next page button
+   * @default false
+   */
+  hideNextButton?: boolean;
+  /**
+   * If true, hides the previous page button
+   * @default false
+   */
+  hidePreviousButton?: boolean;
+  /**
+   * Number of page buttons before and after the current page
+   * @default 1
+   */
+  numberOfAdjacentPages?: number;
+  /**
+   * Number of page buttons at the beginning and end of the page number buttons list
+   * @default 1
+   */
+  numberOfEdgePages?: number;
+  /**
+   * Event that fires when the page number changes
+   */
+  onPageChange?: (event: React.SyntheticEvent, newPage: number) => void;
+  /**
+   * Current page number
+   */
+  page?: number;
+  /**
+   * If true, shows the first page button
+   * @default false
+   */
+  showFirstButton?: boolean;
+  /**
+   * If true, shows the last page button
+   * @default false
+   */
+  showLastButton?: boolean;
+}
+
+export function usePagination(props: Partial<UsePaginationProps> = {}) {
```

Comment:
This hook should be in `src/components/pagination` ... it's the logic for building a pagination component :D even where it's used in TablePagination... that's really just an example of using the hook to create a custom pagination example

`src/hooks` should be nice generic hooks like `useDataPagination` `useControlled` etc... either to manipulate data... or something more at a global app level like media queries or clipboard management... things that stand alone without being the logic to a component 🤔